homeassistant/components/flux_led/__init__.py

```diff
@@ -1 +1,206 @@
-"""The flux_led component."""
+"""The Flux LED/MagicLight integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+import logging
+from typing import Any, Final
+
+from flux_led import BulbScanner, WifiLedBulb
+
+from homeassistant import config_entries
+from homeassistant.components.light import DOMAIN as LIGHT_DOMAIN
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_MODE,
+    CONF_DEVICES,
+    CONF_HOST,
+    CONF_MAC,
+    CONF_MODE,
+    CONF_NAME,
+    CONF_PLATFORM,
+    CONF_PROTOCOL,
+    EVENT_HOMEASSISTANT_STARTED,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.debounce import Debouncer
+from homeassistant.helpers.event import async_track_time_interval
+from homeassistant.helpers.typing import ConfigType
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_COLORS,
+    CONF_CUSTOM_EFFECT,
+    CONF_CUSTOM_EFFECT_COLORS,
+    CONF_CUSTOM_EFFECT_SPEED_PCT,
+    CONF_CUSTOM_EFFECT_TRANSITION,
+    CONF_SPEED_PCT,
+    CONF_TRANSITION,
+    DEFAULT_EFFECT_SPEED,
+    DISCOVER_SCAN_TIMEOUT,
+    DOMAIN,
+    FLUX_HOST,
+    FLUX_LED_EXCEPTIONS,
+    FLUX_MAC,
+    MODE_AUTO,
+    STARTUP_SCAN_TIMEOUT,
+    TRANSITION_GRADUAL,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: Final = ["light"]
+DISCOVERY_INTERVAL: Final = timedelta(minutes=15)
+REQUEST_REFRESH_DELAY: Final = 0.35
+
+
+async def async_wifi_bulb_for_host(hass: HomeAssistant, host: str) -> WifiLedBulb:
+    """Create a WifiLedBulb from a host."""
+    return await hass.async_add_executor_job(WifiLedBulb, host)
+
+
+async def async_discover_devices(
+    hass: HomeAssistant, timeout: int
+) -> list[dict[str, str]]:
+    """Discover ledned devices."""
+
+    def _scan_with_timeout() -> list[dict[str, str]]:
+        scanner = BulbScanner()
+        discovered: list[dict[str, str]] = scanner.scan(timeout=timeout)
+        return discovered
+
+    return await hass.async_add_executor_job(_scan_with_timeout)
+
+
+@callback
+def async_trigger_discovery(
+    hass: HomeAssistant,
+    discovered_devices: list[dict[str, Any]],
+) -> None:
+    """Trigger config flows for discovered devices."""
+    for device in discovered_devices:
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": config_entries.SOURCE_DISCOVERY},
+                data=device,
+            )
+        )
+
+
+@callback
+def async_import_from_yaml(
+    hass: HomeAssistant,
+    config: ConfigType,
+    discovered_mac_by_host: dict[str, str],
+) -> None:
```

Comment:
This isn't ideal because the platform schema is so complex. I think it would be better to do in an `async_setup_platform` in `light.py`, but to make that happen `discovered_mac_by_host` or discovered hosts need to be saved from `async_setup`

Reply:
I'll refactor this a bit after the additional test devices show up

Reply:
Tell me what you're looking for and I can test.

Reply:
I don't have any of the RGBW, RGBWW, or RGBCW devices.

Reply:
I'll test them right now.  

> Since there was a custom effect list per device, it seemed too complex to try to manage that with a single config entry.

The effect are set handled in the front end.  They don't really fit in the configuration as they are something that should be changed whenever you feel like it. 

![image](https://user-images.githubusercontent.com/10890432/135545932-95aca9f1-17ef-469b-9c57-81cf670941f4.png)

Reply:
> I'll test them right now.  
> 
> 
> 
> > Since there was a custom effect list per device, it seemed too complex to try to manage that with a single config entry.
> 
> 
> 
> The effect are set handled in the front end.  They don't really fit in the configuration as they are something that should be changed whenever you feel like it. 
> 
> 
I agree, but I think we should try to avoid breaking users who are using them in that manner at the same time as it's converted to a config flow.  In the future, we could drop it and tell them to use the service.

Reply:
I guess I never thought anyone would use it in the configuration since it's on the LIGHT card.  



> Remaining things to be tested:
> 
> * [ ]  RGBW, RGBWW, or RGBCW devices (I only have RGB)
> * [ ]  A complex yaml config import

I tested RGB, RGBW, and RGBWW.  RGBCW isn't a thing so I didn't test it.  

None of it works well but, it doesn't throw errors or break anything.  Pretty much works the way it did before config flow which was expected.  The bulbs turn on and off. You can set colors.  Color brightness can't be adjusted and neither can the master brightness. I have code ready to fix these but, I need to root around since some things have been moved around. 

I think we should get this PR accepted before adding anything else.  It will fix the issue most users are having.  Which is that the integration doesn't work at all.  At least with this, a user can turn it on and off. 

What do you think?

Reply:
There is something off with this bulb

`129,53,35,97,23,5,255,0,0,0,8,0,240,77,`

Its detecting it as RGBW but its not

Reply:
> There is something off with this bulb
> 
> `129,53,35,97,23,5,255,0,0,0,8,0,240,77,`
> 
> Its detecting it as RGBW but its not

Update flux_led through PyPi and run it.  It's on version 2.3. It may still not work because of they type 35 which is going to be added in over the weekend.  It has better type detection but, I missed the type 35s.

Reply:
updated to 2.3 and still no go

Reply:
If I go into the config options and change it from `auto` to `rgb` it works

Reply:
> If I go into the config options and change it from `auto` to `rgb` it works

Yea, that's what I did to test.

The final goal will be working auto detect.  The flux_led 2.2 would only report if it was RGB capable and how many white channels.  Flux_LED 2.3 actually returns the exact type.  It makes it much easier to auto detect in homeassistant.  

There's two problems.  The first is that current code for the integration does not detect very well and only recognizes the 2.2 method of reporting.  The second problem is that your light is a type 35 which hasn't been added to the list of bulbs that use the new reporting in 2.3. It'll be fixed this weekend since it's a current issue.

Reply:
The bulb says RGBCW on the side of it 🤷

Reply:
One of these showed up and it it works in RGBCW mode

https://www.amazon.com/gp/product/B01DY56N8U/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1

Reply:
```
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 B4E8422CB2B4 (b4:e8:42:2c:b2:b4) with raw_state=bytearray(b'\x813#a#\x01$\xff\x91\x00\t\x00\xf0\t') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 18B905C5239B (18:b9:05:c5:23:9b) with raw_state=bytearray(b'\x813#a#\x10\x00\x00\xff\x00\t\x00\xf0c') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ200 8CCE4E03EE81 (8c:ce:4e:03:ee:81) with raw_state=bytearray(b'\x81%#a\x00\xff\xff\x00\x00\x00\x02\x00\x00*') is rgbww
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 B4E8424401EE (b4:e8:42:44:01:ee) with raw_state=bytearray(b'\x813#a#\x10\x00\x00\xff\x00\t\x00\xf0c') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2145 B4E84214155A (b4:e8:42:14:15:5a) with raw_state=bytearray(b'\x815#a\x17\x05\x00\xff\x00\x00\x08\x00\xf0M') is rgbww
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2146 249494176B09 (24:94:94:17:6b:09) with raw_state=bytearray(b'\x813#a#\x10\x00?\xff\x00\n\x00\xf0\xa3') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2146 249494861B97 (24:94:94:86:1b:97) with raw_state=bytearray(b'\x813#a#\x10\xff\x00\x00\x00\n\x00\xf0d') is rgb
2021-09-30 21:00:13 DEBUG (MainThread) [homeassistant.components.flux_led.light] Detected mode for AK001-ZJ2146 24949416B10D (24:94:94:16:b1:0d) with raw_state=bytearray(b'\x813#a#\x10\x00\x00\xff\x00\n\x00\xf0d') is rgb
```

Reply:
> One of these showed up and it it works in RGBCW mode
> 
> https://www.amazon.com/gp/product/B01DY56N8U/ref=ppx_yo_dt_b_asin_title_o07_s00?ie=UTF8&psc=1
> ```
> % python3 -m flux_led 192.168.214.125 --info
> Unknown ID [192.168.214.125] ON  [Color: (0, 0, 255) White: 0 raw state: 129,37,35,97,0,255,0,0,255,0,2,0,0,42,]
> ```

Yes and you can change the mode.   Those are what use to test the updated detection l.  In the app if you long press the light there's an option to change type.

Reply:
> The bulb says RGBCW on the side of it 🤷 

The type is 35 which is an RGBWW bulb but Flux_LED doesn't have a definition for 35 yet.

The other controllers you have are defined in Flux_LED.

Reply:
I've tested with everything else and it appears to be working as expected.

Reply:
Based on this issue https://github.com/Danielhiversen/flux_led/issues/79 it looks like the same problem.  I'm comfortable its not a regression.  We can work though that after this merges